ambari-server/src/main/java/org/apache/ambari/server/state/cluster/ClusterImpl.java

```diff
@@ -2847,10 +2847,11 @@ private ClusterHealthReport getClusterHealthReport(
     List<Long> hostIds = hosts.stream().map(Host::getHostId).collect(Collectors.toList());
     List<HostComponentDesiredStateEntity> hostComponentDesiredStateEntities =
         hostIds.isEmpty() ? Collections.EMPTY_LIST : hostComponentDesiredStateDAO.findByHostsAndCluster(hostIds, clusterId);
-    Map<Long, Map<String, HostComponentDesiredStateEntity>> mappedHostIds = hostComponentDesiredStateEntities.stream().collect(
+    Map<Long, Map<Long, Map<String, HostComponentDesiredStateEntity>>> mappedHostIds = hostComponentDesiredStateEntities.stream().collect(
+            Collectors.groupingBy(HostComponentDesiredStateEntity::getServiceGroupId,
```

Comment:
Would it be possible to create a single-level map where keys are host component IDs?  That should prevent any duplication problems, and simplifies the loop below.  If not possible, that's OK.  Also, in the description you mention service ID, while the code refers to service group ID, so can you please double-check which one is intended?